lib/Optimizer/CodeGen/QuakeToLLVM.cpp

```diff
@@ -1213,6 +1214,137 @@ class RemoveRelaxSizeRewrite : public OpConversionPattern<quake::RelaxSizeOp> {
+    if (numParameters != 0) {
+      op.emitOpError("Parameterized custom operations not yet supported.");
+    }
+
+    auto arrType = cudaq::opt::getArrayType(context);
+    auto qirArrayTy = cudaq::opt::getArrayType(context);
+    FlatSymbolRefAttr concatFunc =
+        cudaq::opt::factory::createLLVMFunctionSymbol(
+            cudaq::opt::QIRArrayConcatArray, arrType, {arrType, arrType},
+            parentModule);
+
+    // targets
+    auto targetArr = wrapQubitInArray(loc, rewriter, parentModule,
+                                      adaptor.getTargets().front());
+    for (auto oper : adaptor.getTargets().drop_front(1)) {
+      auto backArr = wrapQubitInArray(loc, rewriter, parentModule, oper);
+      auto glue = rewriter.create<LLVM::CallOp>(
+          loc, qirArrayTy, concatFunc, ArrayRef<Value>{targetArr, backArr});
+      targetArr = glue.getResult();
+    }
+
+    // controls
+    auto controls = op.getControls();
+    Value controlArr;
+    if (controls.empty()) {
+      // make an empty array
+      Value zero = rewriter.create<arith::ConstantIntOp>(loc, 0, 64);
+      Value zero32 = rewriter.create<arith::ConstantIntOp>(loc, 8, 32);
+      FlatSymbolRefAttr symbolRef =
+          cudaq::opt::factory::createLLVMFunctionSymbol(
+              cudaq::opt::QIRArrayCreateArray,
+              cudaq::opt::getArrayType(context),
+              {rewriter.getI32Type(), rewriter.getI64Type()}, parentModule);
+      controlArr = rewriter
+                       .create<LLVM::CallOp>(
+                           loc, TypeRange{cudaq::opt::getArrayType(context)},
+                           symbolRef, ValueRange{zero32, zero})
+                       .getResult();
+    } else {
+      controlArr = wrapQubitInArray(loc, rewriter, parentModule,
+                                    adaptor.getControls().front());
+      for (auto oper : adaptor.getControls().drop_front(1)) {
+        auto backArr = wrapQubitInArray(loc, rewriter, parentModule, oper);
+        auto glue = rewriter.create<LLVM::CallOp>(
+            loc, qirArrayTy, concatFunc, ArrayRef<Value>{controlArr, backArr});
+        controlArr = glue.getResult();
+      }
+    }
+    // Fetch the unitary matrix generator for this custom operation
+    auto sref = op.getGenerator();
+    StringRef generatorName = sref.getRootReference();
+    auto globalOp =
+        parentModule.lookupSymbol<cudaq::cc::GlobalOp>(generatorName);
+    if (globalOp) {
+      auto complex64Ty =
+          typeConverter->convertType(ComplexType::get(rewriter.getF64Type()));
+      auto complex64PtrTy = LLVM::LLVMPointerType::get(complex64Ty);
+      Type type = getTypeConverter()->convertType(globalOp.getType());
+      auto addrOp =
+          rewriter.create<LLVM::AddressOfOp>(loc, type, generatorName);
+      auto unitaryData =
+          rewriter.create<LLVM::BitcastOp>(loc, complex64PtrTy, addrOp);
+
+      auto qirFunctionName = "__quantum__qis__custom_unitary";
```

Comment:
These qis function names should be put in cudaq/Optimizer/CodeGen/QIRFunctionNames.h